File: packages/vision_ai/ios/Classes/Core/FrameProcessor.swift
import AVFoundation
import CoreVideo

// Orchestrates hand + face ML processing for each camera frame.
// Called on the analysis queue by CameraManager's AVCaptureVideoDataOutputSampleBufferDelegate.
// Hand processing is async (MediaPipe LIVE_STREAM); face processing is synchronous (ML Kit).
class FrameProcessor {

    let resultAggregator: ResultAggregator
    var handProcessor: HandGestureProcessor?
    var faceProcessor: FaceDetectionProcessor?
    let pixelBufferPool = PixelBufferPool() // shared across processors to reduce allocations
    private var frameCount: Int64 = 0

    init(
        resultAggregator: ResultAggregator,
        handProcessor: HandGestureProcessor?,
        faceProcessor: FaceDetectionProcessor?
    ) {
        self.resultAggregator = resultAggregator
        self.handProcessor = handProcessor
        self.faceProcessor = faceProcessor
    }

    // Called on analysis queue for each camera frame
    func processFrame(pixelBuffer: CVPixelBuffer, isFrontCamera: Bool) {
        let startTime = ProcessInfo.processInfo.systemUptime
        let timestampMs = Int(startTime * 1000)

        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)

        frameCount += 1
        if frameCount % 100 == 0 {
            print("VisionAI.FrameProcessor: Processed \(frameCount) frames")
        }

        // Hand: async — feed frame to MediaPipe, pick up latest result
        if let hp = handProcessor {
            hp.processFrame(pixelBuffer: pixelBuffer, timestampMs: timestampMs)
        }

        // Face: synchronous — blocks until ML Kit returns (same as Android's Tasks.await)
        var faceResult: FaceProcessorResult?
        if let fp = faceProcessor {
            faceResult = fp.processFrame(pixelBuffer: pixelBuffer, pool: pixelBufferPool)
        }

        // Collect hand results (may be from previous frame due to async)
        let handResult = handProcessor?.getLatestResult()
        let handMaps = handResult?.toMapList() ?? []
        let faceMaps = faceResult?.toMapList() ?? []

        let inferenceTimeMs = Int((ProcessInfo.processInfo.systemUptime - startTime) * 1000)

        resultAggregator.emit(
            hands: handMaps,
            faces: faceMaps,
            imageWidth: width,
            imageHeight: height,
            inferenceTimeMs: inferenceTimeMs
        )
    }

    func release() {
        pixelBufferPool.release()
    }
}

File: packages/vision_ai/android/src/main/kotlin/com/visionai/vision_ai/core/FrameProcessor.kt
package com.visionai.vision_ai.core

import android.graphics.Bitmap
import android.os.SystemClock
import android.util.Log
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.visionai.vision_ai.face.FaceDetectionProcessor
import com.visionai.vision_ai.hand.HandGestureProcessor

class FrameProcessor(
    private val resultAggregator: ResultAggregator,
    private val handProcessor: HandGestureProcessor? = null,
    private val faceProcessor: FaceDetectionProcessor? = null,
    private val isFrontCamera: Boolean = true,
) : ImageAnalysis.Analyzer {

    private var frameCount = 0L

    override fun analyze(imageProxy: ImageProxy) {
        try {
            val startTime = SystemClock.uptimeMillis()
            val timestamp = SystemClock.uptimeMillis()

            frameCount++
            if (frameCount % 100 == 0L) {
                Log.d(TAG, "Processed $frameCount frames")
            }

            // Convert frame once, reuse for both processors
            var bitmap: Bitmap? = null
            if (handProcessor != null || faceProcessor != null) {
                bitmap = ImageConverter.imageProxyToBitmap(imageProxy, isFrontCamera)
            }

            // Hand gesture detection (async via MediaPipe LIVE_STREAM)
            if (handProcessor != null && bitmap != null) {
                handProcessor.processFrame(bitmap, timestamp)
            }

            // Face emotion detection (sync via ML Kit)
            val faceResult = if (faceProcessor != null && bitmap != null) {
                faceProcessor.processFrame(bitmap, imageProxy.imageInfo.rotationDegrees)
            } else {
                null
            }

            val handResult = handProcessor?.getLatestResult()

            val inferenceTime = SystemClock.uptimeMillis() - startTime

            val imageWidth: Int
            val imageHeight: Int
            val rotation = imageProxy.imageInfo.rotationDegrees
            if (rotation == 90 || rotation == 270) {
                imageWidth = imageProxy.height
                imageHeight = imageProxy.width
            } else {
                imageWidth = imageProxy.width
                imageHeight = imageProxy.height
            }

            resultAggregator.emit(
                handResults = handResult?.toMapList() ?: emptyList(),
                faceResults = faceResult?.toMapList() ?: emptyList(),
                timestampMs = timestamp,
                inferenceTimeMs = inferenceTime,
                imageWidth = imageWidth,
                imageHeight = imageHeight,
            )
        } catch (e: Exception) {
            Log.e(TAG, "Frame processing error", e)
        } finally {
            imageProxy.close()
        }
    }

    companion object {
        private const val TAG = "VisionAI.FrameProcessor"
    }
}
